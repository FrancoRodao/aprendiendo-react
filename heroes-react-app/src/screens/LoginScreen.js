import React from 'react'
import { AuthContext } from '../auth/authContext'
import { types } from '../types/types'
import { useContext } from 'react'

export default function LoginScreen( {history} ) {

    const { dispatch } = useContext(AuthContext)

    const handleLogin = ()=>{

        const lastPath = localStorage.getItem("lastPath") ||'/'

        dispatch({
            type: types.login,
            payload: {
                name: "Franco"
            }
        })
        history.push(lastPath)
    }

    return (
        <div className="container mt-5">
            <h2>login screen</h2>
            <hr/>

            <button
                className="btn btn-primary"
                onClick={handleLogin}
            >
                login
            </button>
        </div>
    )
}
