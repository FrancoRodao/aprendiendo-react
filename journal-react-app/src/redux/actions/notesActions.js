import { db } from "../../firebase/firebase.config"
import { imageUploadToCloudinary } from "../../helpers/imageUpload"
import { loadNotes } from "../../helpers/loadNotes"
import { types } from "../types/types"

export const startAddNewNote = () => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        const newNote = {
            title: 'title',
            body: 'body',
            date: new Date().getTime()
        }

        const noteRef = await db.collection(`${userId}/journal/notes`).add(newNote)
        dispatch(addNewNote(noteRef.id, newNote))
        dispatch(activeNoteAction(noteRef.id, newNote))
    }
}

export const addNewNote = (noteId, note) => ({
    type: types.notesAdd,
    payload: {
        id: noteId,
        note
    }
})

export const activeNoteAction = (id, note) => ({
    type: types.notesSetActive,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (userId) => {
    return async (dispatch) => {
        const notes = await loadNotes(userId)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth
        const noteToFireStore = {...note}
        delete noteToFireStore.id

        await db.collection(`${uid}/journal/notes`).doc(`${note.id}`).update(noteToFireStore)
        dispatch(refreshNote(note))
    }
}

export const refreshNote = (note) => ({
    type: types.notesUpdate,
    payload: {
        note
    }
})

export const startUploadPicture = (file) => {
    return async (dispatch, getState) => {
        const activeNote = getState().notes.active
        const imageUrl = await imageUploadToCloudinary(file)
        activeNote.imgUrl = imageUrl
        
        dispatch(startSaveNote(activeNote))
    }
}


export const startDeleteNote = (note) => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        await db.doc(`${userId}/journal/notes/${note.id}`).delete()
        dispatch(deleteNote(note))
    }
}

export const deleteNote = (note) => ({
    type: types.notesDelete,
    payload: {
        note
    }
})