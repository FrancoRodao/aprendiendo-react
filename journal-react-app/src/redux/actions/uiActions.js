import { types } from "../types/types"

export const startLoadingAction = () => (
    {
        type: types.uiStartLoading,
        payload: {
            loading: true
        }
    }
)


export const finishLoadingAction = () => (
    {
        type: types.uiFinishLoading,
        payload: {
            loading: false
        }
    }
)

export const errorAction = (state, message) => (
    {
        type: types.uiError,
        payload: {
            state,
            lastMsg: message
        }
    }
)

export const clearErrorsAction = () =>  (
    {
        type: types.uiError,
        payload: {
            state: false,
            lastMsg: ""
        }
    }
)