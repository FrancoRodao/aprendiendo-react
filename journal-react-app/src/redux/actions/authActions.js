import { types } from "../types/types"
import { firebase, googleAuthProvider } from "../../firebase/firebase.config";
import { clearErrorsAction, errorAction, finishLoadingAction, startLoadingAction } from "./uiActions";

export const startLogin = (email, password) => {
    return (dispatch) => {
        dispatch(startLoadingAction())

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(e => {
                if (e.code === "auth/user-not-found" || "auth/wrong-password") {
                    dispatch(errorAction(true, "Wrong credentials"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const startGoogleLogin = () => {
    return dispatch => {
        dispatch(startLoadingAction())

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const loginAction = (uid, displayname) => (
    {
        type: types.authLogin,
        payload: {
            uid,
            displayname
        }
    }
)

export const startRegisterUser = (name, email, password) => {
    return (dispatch, state) => {
        dispatch(startLoadingAction())
        return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name,
                    photoURL: ''
                })
                dispatch(registerUserAction(user.uid, user.displayName))
                dispatch( clearErrorsAction() )
            })
            .catch(e => {
                if (e.code === "auth/email-already-in-use") {
                    dispatch(errorAction(true, "Email already registered"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .then(() => state())
            .finally(() => dispatch(finishLoadingAction()))
        
    }
}

export const registerUserAction = (uid, displayname) => {
    return {
        type: types.register,
        payload: {
            uid,
            displayname
        }
    }
}


export const startLogout = () => {
    return async (dispatch) => {
        dispatch(startLoadingAction())

        firebase.auth().signOut()
            .then(() => dispatch(logoutUserAction()))
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const logoutUserAction = () => ({
    type: types.authLogout,
    payload: {}
})

