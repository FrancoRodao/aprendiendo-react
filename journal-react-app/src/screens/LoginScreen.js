import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from "validator";
import { startGoogleLogin, startLogin } from '../redux/actions/authActions'
import { errorAction } from '../redux/actions/uiActions';

export default function LoginScreen() {

    const [formErrors, setFormErrors] = useState({
        email: '',
        password: ''
    })

    const dispatch = useDispatch()
    const state = useSelector(state => state)
    const loading = state.ui.loading
    const { state: errorState, lastMsg } = state.ui.errors

    const handleSubmit = (e) => {
        e.preventDefault()
        const form = {
            email: e.target[0].value,
            password: e.target[1].value
        }

        if (isFormValid(form)) {
            dispatch(startLogin(form.email, form.password))
        }
    }

    const isFormValid = ({ email, password }) => {
        let formErrors = {
            email: '',
            password: ''
        }

        if (!validator.isEmail(email)) {
            formErrors.email = "Email is not valid"
        }

        if (password.length === 0) {
            formErrors.password = "Password is required"
        }

        setFormErrors(formErrors)

        if (formErrors.email.length > 0 || formErrors.password.length > 0) {
            return false
        } else {
            return true
        }
    }

    const handleGoogleLogin = () => dispatch(startGoogleLogin())

    //clear errors when change page
    const handleLinkClick = () => dispatch(errorAction(false, ""))

    return (
        <>
            <h3 className={`auth__title ${errorState ? 'auth__title-errorUI':''}`}>
                Login
            </h3>

            {
                errorState && 
                <div className="auth__alert-error-ui">
                    {lastMsg}
                </div>
            }

            <form onSubmit={handleSubmit}>
                <input
                    type="email"
                    placeholder="Email"
                    autoComplete="on"
                    className={
                        `auth__input ${formErrors.email.length > 0 ?
                            'auth__input-error' : ''}`
                    }
                />

                {
                    formErrors.email.length > 0 &&
                    <p className="auth__alert-error">
                        {formErrors.email}
                    </p>
                }

                <input
                    type="password"
                    placeholder="Password"
                    className={
                        `auth__input ${formErrors.password.length > 0 ?
                            'auth__input-error' : ''}`
                    }
                />

                {
                    formErrors.password.length > 0 &&
                    <p className="auth__alert-error">
                        {formErrors.password}
                    </p>
                }

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={loading}
                >
                    {
                        loading ?
                            <i className="fas fa-spinner fa-spin"></i>
                            :
                            "Login"
                    }
                </button>

                <hr></hr>
                <div className="auth__social-networks">
                    <p>Login with social networks</p>

                    <div
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>


                <Link onClick={handleLinkClick} className="link" to="/register" >
                    Create new account
                </Link>

            </form>
        </>
    )
}
