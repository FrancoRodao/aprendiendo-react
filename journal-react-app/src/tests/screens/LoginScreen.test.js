import React from 'react'
import { mount } from "enzyme"
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import LoginScreen from "../../screens/LoginScreen"
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router';
import { startGoogleLogin, startLogin } from '../../redux/actions/authActions';

jest.mock('../../redux/actions/authActions', () => {
    return {
        __esModule: true,
        startGoogleLogin: jest.fn(),
        startLogin: jest.fn()
    }
})

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const storeInitialState = {
    auth: {},
    notes: {},
    ui: {
        loading: false,
        errors: {
            state: false,
            msgErr: ''
        }
    }
}
let store = mockStore(storeInitialState)
store.dispatch = jest.fn()

describe('login screen tests', () => {

    beforeEach(() => store = mockStore(storeInitialState))

    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <LoginScreen />
            </MemoryRouter>
        </Provider>
    )

    test('should match to snapshot', () => {

        expect(wrapper).toMatchSnapshot()
    })

    test('should dispatch google login action', () => {
        wrapper.find('.google-btn').prop('onClick')()
        expect(startGoogleLogin).toHaveBeenCalled()
    })

    test('should dispatch normal login', () => {
        wrapper.find('.btn.btn-primary.btn-block').simulate("submit", {
            preventDefault: () => {},
            target: [
                {
                    value: process.env.REACT_APP_testEmailUser
                },
                {
                    value: process.env.REACT_APP_testPasswordUser
                }
            ]
        })

        expect(startLogin).toHaveBeenCalledWith(process.env.REACT_APP_testEmailUser, process.env.REACT_APP_testPasswordUser)
    })



})
