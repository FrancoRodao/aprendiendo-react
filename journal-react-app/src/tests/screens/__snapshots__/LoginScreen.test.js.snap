// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login screen tests should match to snapshot 1`] = `
Array [
  <h3
    className="auth__title "
  >
    Login
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="on"
      className="auth__input "
      placeholder="Email"
      type="email"
    />
    <input
      className="auth__input "
      placeholder="Password"
      type="password"
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <hr />
    <div
      className="auth__social-networks"
    >
      <p>
        Login with social networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/register"
      onClick={[Function]}
    >
      Create new account
    </a>
  </form>,
]
`;
